### PR TITLE
Add syntax for tables

### PR DESCRIPTION
**Pull request by gpetiot**
This is a first draft, I'm asking for feedback before going further. @jonludlam @Julow

I chose this syntax:

```
{t
  {tr
     {td ...}
     ...
     {td ...}
  }
  ...
  {tr
     {td ...}
     ...
     {td ...}
  }
}
```
The markup is ovsiously inspired from the `<tr></tr><td></td>` HTML tags.
The generated output is straightforward in HTML and Latex, at first I wanted to add header cells (`<th></th>`) but I'm not aware of anything similar in Latex but we could apply a different style.

Some improvements:
- add the ability to say that a cell will span over N>1 columns. In HTML it would be `<td colspan="N">...</td>`, in Latex it would be `\multicolumn{N}{c}{...}`
- check that each row has the same number of columns? (I don't remember if that will cause issues for Latex, but HTML should be more lenient).

What do you think?

### CODE REVIEW

**Comment by jonludlam** [comment]
Thanks @gpetiot !

This is definitely an approach that could work very well. However, I think we ought to consider the possibility of having more visual way to layout tables, perhaps something similar to how rst or orgmode does it - as an example both of the following are valid rst tables:

```
+------------+------------+-----------+
| Header 1   | Header 2   | Header 3  |
+============+============+===========+
| body row 1 | column 2   | column 3  |
+------------+------------+-----------+
| body row 2 | Cells may span columns.|
+------------+------------+-----------+
| body row 3 | Cells may  | - Cells   |
+------------+ span rows. | - contain |
| body row 4 |            | - blocks. |
+------------+------------+-----------+
```

and 

```
=====  =====  ======
   Inputs     Output
------------  ------
  A      B    A or B
=====  =====  ======
False  False  False
True   False  True
False  True   True
True   True   True
=====  =====  ======
```

and here's an org-mode table:

```
| N | N^2 | N^3 | N^4 | sqrt(n) | sqrt[4](N) |
|---+-----+-----+-----+---------+------------|
| / |  <  |     |  >  |       < |          > |
| 1 |  1  |  1  |  1  |       1 |          1 |
| 2 |  4  |  8  | 16  |  1.4142 |     1.1892 |
| 3 |  9  | 27  | 81  |  1.7321 |     1.3161 |
|---+-----+-----+-----+---------+------------|
```


The reason being that more often than not right now I'm still looking at docs in my editor buffer, or on github, as opposed to nicely rendered on a web page. We'd also get to use pre-existing table-editing modes in various editors.

**Comment by Julow** [comment]
Features that other languages have: text alignment of the column, header rows.

The short syntax from RST is interesting: https://docutils.sourceforge.io/docs/ref/rst/directives.html#list-table if we drop the structured directive and attributes and make it less ambiguous, we get a syntax easy to type that is similar to lists and might be enough in most cases.

I suggest this syntax, which works the same way as lists:

```
++ - header 1
   - header 2
** - cell 1
   - cell 2
```

The long syntax that you suggest is needed, like we have for lists, to support nesting of bigger elements.

To leave room for the `colspan` feature and text alignment, we need some kind of attributes: `{td[colspan=2,align=center] ...}` (similar to CSS selectors).

Header row similar to HTML:
```
{table
 {tr
  {th header 1}
  {th header 2}}
 {tr
  {td cell 1}
  {td cell 2}}}
```

The main element can be `{table ...}`, it's more readable and similar to HTML.

**Comment by dbuenzli** [comment]
Most lightweight markup languages add two things, headers and specification of column alignement (see for example the github flavoured markdown [table definition](https://github.github.com/gfm/#tables-extension-)). I think it would be nice to have these. I wouldn't bother too much with multi-columns.

I have mixed feelings about what @jonludlam suggests. Of course these tables look nicer but the given examples are also not very content heavy.

Once you start adding links, references and ocamldoc markup in the cell contents I have the impression that the regular, well-delimited, syntax of ocamldoc scales better.

**Comment by Julow** [comment]
> The reason being that more often than not right now I'm still looking at docs in my editor buffer, or on github, as opposed to nicely rendered on a web page.

I've always hated both reading and writing tables drawn and aligned like that. The editor plugins to draw them are not worth learning for the few tables I ever needed to make so I end up not aligning them, they become write-only. Also, multi line cells are a pain.

**Comment by jonludlam** [comment]
@dbuenzli agreed that larger-scale tables won't look as nice as those examples! Nothing stopping us from doing both though - we already have precedent in lists, where we can write both
```
(** Here is a {b list}
- item 1
- item 2
- item 3

The list is ended by the blank line.*)

```
and
```
(** Here is a {b list}
{ul {- item 1}
{- item 2}
{- item 3}}
The list is ended by the blank line.*)
```

we could simply choose based on the content of the `{table ...}` element - e.g.

```
{table
 {tr
  {th header 1}
  {th header 2}}
 {tr
  {td cell 1}
  {td cell 2}}}
```

vs

```
{table
+==========+==========+
| header 1 | header 2 |
+==========+==========+
| cell 1   | cell 2   |
+----------+----------+
}
```

**Comment by jonludlam** [comment]
> I've always hated both reading and writing tables drawn and aligned like that. The editor plugins to draw them are not worth learning for the few tables I ever needed to make so I end up not aligning them, they become write-only. Also, multi line cells are a pain.

I've never hated _reading_ them, but writing is a bit of a pain without editor support :-)

**Comment by jonludlam** [comment]
Actually, choosing based on contents is not a good idea. We could use a different tag though - `{t` vs `{table` for example.

**Comment by dbuenzli** [comment]
To motivate with concrete examples here's a sample of tables that I currently write as code and that would benefit from the proposal:

1. https://erratique.ch/software/b0/doc/B00_ocaml/Mod/Name/index.html#mangled
2. https://erratique.ch/software/brr/doc/ffi_cookbook.html#names
3. https://erratique.ch/software/brr/doc/ffi_manual.html#equality
4. https://erratique.ch/software/gg/doc/Gg/Float/index.html#floatrecall
5. https://erratique.ch/software/uutf/doc/Uutf/index.html#val-decoder 
6. https://erratique.ch/repos/inane/tree/src/b36uid.mli#n12

Scroll a bit if you don't see the table, which reminds me: it would be nice to be able to reference them via a label like we do for sections, i.e. `{table:label` (and why not cells and/or rows, but let's start simple).

**Comment by dbuenzli** [comment]
Regarding a possible lightweight notation. I think it's good at least have a look at what the [experts do](https://htmlpreview.github.io/?https://github.com/jgm/djot/blob/master/doc/syntax.html#pipe-table) to when they can start designing from scratch.

**Comment by gpetiot** [comment]
For the light syntax, what do you think about something like:

```
{t
+-----------------------------
|{l}  xx |{c}  yy  |{r}   zz
+---------------------------
|   xx   |{r}   yy  |{l}    zz
+-----------------------------
}
```

- No need to finish the line (just `+-` at the beginning of the line to indicate a new row)
- `{l}`/`{r}`/`{c}` could indicate the alignment (optional), but should be stuck to the `|` to be a special token

**Comment by dbuenzli** [comment]
> For the light syntax, what do you think about something like:

If you really want to add this then I strongly suggest not to invent anything new here. Have a look at [pandoc's markdown table extensions](https://pandoc.org/MANUAL.html#tables) for a few alternatives. 

Personally I think you should just go with `pipe_tables` which is popular among a few extended markdown dialects including [github's](https://github.github.com/gfm/#tables-extension-) and [gitlab's](https://docs.gitlab.com/ee/user/markdown.html#tables) ones.

**Comment by gpetiot** [comment]
I implemented the light and heavy syntaxes, this PR is ready to be reviewed.

File: src/syntax.ml [hunk 1]
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
...
+      |> add_warning input;
+      ([], [], [], brace_location)
+  | [] :: _ ->
+      Parse_error.not_allowed ~what:"missing header"
+        ~in_what:(Token.describe parent_markup)
+        brace_location
+      |> add_warning input;
+      ([], [], [], brace_location)
+  | header :: [] ->
+      Parse_error.not_allowed ~what:"missing separator between header and data"
```
Comment (Julow) [inline src/syntax.ml]: This concatenates "is not allowed", which is not very nice. It should be a new kind of error.

File: src/syntax.ml [hunk 2]
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
...
+        ~in_what:(Token.describe parent_markup)
+        brace_location
+      |> add_warning input;
+      (header, [], [], brace_location)
+  | header :: align :: data ->
```
Comment (Julow) [inline src/syntax.ml]: This is arbitrary. I want tables without a header.

File: src/syntax.ml [hunk 3]
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
...
+        let w =
+          if String.get w (String.length w - 1) = '|' then
+            String.sub w 0 (String.length w - 1)
+          else w
+        in
+        let ws = String.split_on_char '|' w in
+        let ws = List.map (fun x -> { Loc.value = `Word x; location }) ws in
+        if List.for_all (fun x -> Option.is_some (valid_align x)) ws then
+          (* revert the list for consistency *)
+          Some (List.map (fun x -> [ x ]) ws |> List.rev)
```
Comment (Julow) [inline src/syntax.ml]: The extra wrapping in a list adds some "impossible" cases later. I'd prefer if the separator row had its own more comfortable type. It doesn't need to be mixed with the other rows later, the caller code needs to be changed.

Same for the reversed list, it feels wrong.

File: src/syntax.ml [hunk 4]
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
...
+          if String.get w 0 = '|' then String.sub w 1 (String.length w - 1)
+          else w
+        in
+        let w =
+          if String.get w (String.length w - 1) = '|' then
+            String.sub w 0 (String.length w - 1)
+          else w
+        in
+        let ws = String.split_on_char '|' w in
+        let ws = List.map (fun x -> { Loc.value = `Word x; location }) ws in
```
Comment (Julow) [inline src/syntax.ml]: `valid_align` doesn't need to work on `` `Word ``, removing a layer of match and simplifying the `List.map`s.

File: src/syntax.ml [hunk 5]
```diff
@@ -1067,6 +1149,354 @@ and explicit_list_items :
...
+        |> add_warning input;
+        junk input;
+        (acc_t, acc_r, acc_c, last_loc)
+  in
+  let acc_t, acc_r, acc_c, brace_location =
+    consume_elements [] [] [] ~new_line:true ~last_loc:parent_markup_location
+  in
+  let acc_r = match acc_c with [] -> acc_r | _ -> List.rev acc_c :: acc_r in
+  let acc_t = match acc_r with [] -> acc_t | _ -> List.rev acc_r :: acc_t in
+  let rows = List.rev acc_t in
```
Comment (Julow) [inline src/syntax.ml]: `acc_t`, `acc_r`, `acc_c` that's too hard to follow and adds unnecessary constraints: separator row validated once then parsed again, "impossible" cases, duplicated code.

I suggest making two "consume" functions: One for the header and separator, the other for the remaining rows. That way, the parsing of the first two is not mentioned all over the place and can return the more interesting type directly.
I believe this can also help to support an optional header row, and remove the duplicated error checking.

**Review by Julow — COMMENTED** [review]
This parses as two columns, `a` and `b |c| d`, which is not intuitive:

```
{t
 | a | b |c| d |
 |---|--:|:--|:-:|
}
```

**Comment by jonludlam** [comment]
I assume that, since pipetables are pretty standard, there are a set of nasty test cases somewhere that we could use?

**Comment by dbuenzli** [comment]
Would it please be possible to update the first comment with the implemented syntaxes. For example how do I specify alignement in the "heavy" syntax ?

File: src/syntax.ml (hunk 1 above)
Comment (gpetiot) [reply src/syntax.ml]: I rewrote this pattern matching a, more things are allowed now (no header, no align row)

File: src/syntax.ml (hunk 2 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/syntax.ml (hunk 5 above)
Comment (gpetiot) [reply src/syntax.ml]: I've split the consume function in two. But if we would change the design to first expect the header, then the align row, then the data it would not allow to have tables containing a single line, or only data, etc. I find it convenient to have the align row have the same type as the data (represented as Word) and interpret it when we parsed everything (allow more flexible tables). The code has been heavily modified, what do you think?

**Comment by gpetiot** [comment]
> Would it please be possible to update the first comment with the implemented syntaxes. For example how do I specify alignement in the "heavy" syntax ?

Nothing decided yet, I'm torn between doing something similar to the light syntax:
```
{table
   {tr {th ...} [th ...}}
   {align left right}
   ...
}
```
or something that would be easier to read to my opinion but that would make both implementations diverge:
```
{table
  {tr {th{left} ...} {th{right} ...}} or {tr {th:left ...} {th:right ...}}
  ...
}
```

**Comment by dbuenzli** [comment]
Note that even though IIRC `odoc` dropped support for it. There was a syntax in `ocamldoc` for [text alignement](https://v2.ocaml.org/manual/ocamldoc.html#inline-text-element). Might be worth resurecting it for `th`. Something like: a bare `th` is left-aligned (or whathever the default is of the light syntax) and otherwise alignement is specified by "direct" nesting:

```
{tr
{th {L Left aligned}} {th {R Right aligned}} {th {C Center aligned}} {th default align}}
```

**Comment by gpetiot** [comment]
> Note that even though IIRC `odoc` dropped support for it. There was a syntax in `ocamldoc` for [text alignement](https://v2.ocaml.org/manual/ocamldoc.html#inline-text-element). Might be worth resurecting it for `th`. Something like: a bare `th` is left-aligned (or whathever the default is of the light syntax) and otherwise alignement is specified by "direct" nesting:
> 
> ```
> {tr
> {th {L Left aligned}} {th {R Right aligned}} {th {C Center aligned}} {th default align}}
> ```

Sounds good to me, @jonludlam what do you think?

**Comment by jonludlam** [comment]
Reusing the syntax we previously had makes sense to me.

**Comment by Julow** [comment]
On the other hand, an attribute-like syntax would allow `colspan` and keep room for new things later.

File: src/syntax.ml (hunk 3 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/syntax.ml (hunk 4 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/compat.ml [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  type 'a t = 'a option = None | Some of 'a
+
+  let is_some = function None -> false | Some _ -> true
+
+  let value_exn = function
+    | None -> failwith "Option.value_exn None"
+    | Some x -> x
+
+  let join_list l =
+    if List.for_all Option.is_some l then Some (List.map value_exn l) else None
```
Comment (Julow) [inline src/compat.ml]: Using the stdlib's module, should be:
```ocaml
    if List.for_all is_some l then Some (List.map value_exn l) else None
```

File: src/syntax.ml [hunk 7]
```diff
@@ -1067,6 +1153,358 @@ and explicit_list_items :
...
+    * Ast.inline_element with_location list list list
+    * Loc.span =
+ fun ~parent_markup ~parent_markup_location input ->
+  let valid_align = function
+    | [ { Loc.value = `Word ""; _ } ] -> Some `Default
+    | [ { Loc.value = `Word w; _ } ] ->
+        if String.length w = 1 then
+          match w with "-" -> Some `Default | ":" -> Some `Center | _ -> None
+        else if
+          String.for_all (Char.equal '-') (String.sub w 1 (String.length w - 2))
```
Comment (Julow) [inline src/syntax.ml]: This is `String.contains (...) '-'`. Two compat functions can be removed.

This function would be easier to read if it was broken into smaller functions. Also, `valid_align` could benefit from some local variables to avoid copying the string or to store the length of string.

File: src/compat.ml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+end
+
+module Char = struct
+  include Char
+
+  let equal x y = compare x y = 0
```
Comment (Julow) [inline src/compat.ml]: This is the slow polymorphic version of the function. I'd prefer if this wasn't defined. The better version is:
```ocaml
let equal (x : char) y = x = y
```

File: src/compat.ml [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  let for_all f str =
+    let rec aux i =
+      if i >= String.length str then true
+      else if f (String.get str i) then aux (i + 1)
+      else false
+    in
+    aux 0
+
+  let split_on_char sep s =
```
Comment (Julow) [inline src/compat.ml]: It doesn't seem to be used.

**Review by Julow — COMMENTED** [review]
I'm not a fan of the pipe syntax at all but using the most common syntax can't hurt Odoc.

File: src/syntax.ml (hunk 7 above)
Comment (gpetiot) [reply src/syntax.ml]: But I want to check that it only contains '-'.

**Comment by gpetiot** [comment]
I fixed the parsing issues with the alignment tags for the heavy syntax. Let me know what you think @Julow @jonludlam

File: src/syntax.ml [hunk 10]
```diff
@@ -404,6 +406,8 @@ type where_in_line =
   | `After_tag
   | `After_shorthand_bullet
   | `After_explicit_list_bullet
+  | `After_table_header
+  | `After_table_cell
```
Comment (panglesd) [inline src/syntax.ml]: Maybe the comment above could be updated!

File: src/syntax.ml [hunk 11]
```diff
@@ -561,6 +569,8 @@ let rec block_element_list :
         | Top_level -> (List.rev acc, next_token, where_in_line)
         | In_shorthand_list -> (List.rev acc, next_token, where_in_line)
         | In_explicit_list -> (List.rev acc, next_token, where_in_line)
+        | In_table_header -> (List.rev acc, next_token, where_in_line)
+        | In_table_cell -> (List.rev acc, next_token, where_in_line)
```
Comment (panglesd) [inline src/syntax.ml]: Maybe also update the comment above the `block_element_list` function definition?

File: src/ast.ml [hunk 12]
```diff
@@ -41,13 +52,18 @@ type nestable_block_element =
...
   | `Math_block of string  (** @since 2.0.0 *) ]
 (** Some block elements may be nested within lists or tags, but not all.
     The [`List] constructor has a parameter of type [\[`Light | `Heavy\]].
     This corresponds to the syntactic constructor used (see the
     {{:https://ocaml.org/releases/4.12/htmlman/ocamldoc.html#sss:ocamldoc-list}manual}).
     *)
 
+and table =
+  | Light of inline_element abstract_table
+  | Heavy of nestable_block_element abstract_table
```
Comment (Julow) [inline src/ast.ml]: The ast is entirely made of structural types, this should be a polymorphic variant for consistency. Same for `abstract_table` which could be a simple tuple.

File: src/syntax.ml [hunk 13]
```diff
@@ -1067,6 +1139,330 @@ and explicit_list_items :
...
+        Parse_error.not_allowed next_token.location ~what:(Token.describe token)
+          ~in_what:(Token.describe parent_markup)
+          ~suggestion
+        |> add_warning input;
+        junk input;
+        consume_cell_items acc
+  in
+  consume_cell_items []
+
+and heavy_table_header ~parent_markup input =
```
Comment (Julow) [inline src/syntax.ml]: The 3 functions have very similar code: Parsing nested-block-elements until a `` `Right_brace ``. The only change is how to handle `` `Begin_table_header | `Begin_table_data | `Begin_paragraph_style `` etc.. I think a more general higher order function is possible.

**Review by Julow — COMMENTED** [review]
Nice!

Adding a block element into a light table makes quite a mess of warnings and continue the parsing in an inconsistent way:
```ocaml
# p {|
  {t
   | {[ a ]} | b |
   |---|---|
  }
  |} ;;
File "", line 2, characters 5-12:
'{[...]}' (code block) is not allowed in '{t ...}' (table).
File "", line 2, characters 13-14:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 2, characters 15-16:
Paragraph should begin on its own line.
File "", line 2, characters 17-18:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 3-4:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 7-8:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 3, characters 8-11:
Paragraph should begin on its own line.
File "", line 3, characters 11-12:
'|' is not allowed in top-level text.
Suggestion: move '|' into '{t ...}' (table).
File "", line 4, characters 2-3:
Unpaired '}' (end of markup).
Suggestion: try '\}'.
- : Ast.t =
[`Table (Light {header = []; data = []; align = []});
 `Paragraph [`Word "b"; `Space " "]; `Paragraph [`Word "---"];
 `Paragraph [`Word "---"]; `Paragraph [`Word "}"]]
```


The parser has the goal of not loosing text in case of bad syntax so the problematic block element shouldn't be removed. It would be nice to handle the conversion from table to paragraphs without generating too much warnings.

File: src/ast.ml (hunk 12 above)
Comment (gpetiot) [reply src/ast.ml]: done

File: src/syntax.ml (hunk 13 above)
Comment (gpetiot) [reply src/syntax.ml]: I factorized a bit and removed some unnecessary warnings when parsing.

File: src/syntax.ml (hunk 11 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/syntax.ml (hunk 10 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/syntax.ml [hunk 14]
```diff
@@ -36,6 +38,107 @@ let peek input =
...
+          `End (acc, next_token.location)
+      | `Space _ | `Single_newline _ | `Blank_line _ ->
+          junk input;
+          consume ()
+      | _ -> `Token next_token
+    in
+    consume ()
+
+  module Infix = struct
+    let ( let>> ) consume if_token =
```
Comment (panglesd) [inline src/syntax.ml]: this syntax is only available with ocaml 4.08. `odoc-parser` should be available starting from 4.02.

```suggestion
    let ( >>> ) consume if_token =
```

File: src/syntax.ml [hunk 15]
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
...
   consume_list_items []
 
+(* Consumes a sequence of table rows that might start with [`Bar].
+
+   This function is called immediately after '{t' ([`Begin_table `Light]) is
+   read. The only "valid" way to exit is by reading a [`Right_brace] token,
+   which is consumed. *)
+and light_table ~parent_markup ~parent_markup_location input =
+  let rec consume_rows acc ~last_loc =
+    let>> next_token = Reader.until_rbrace input acc in
```
Comment (panglesd) [inline src/syntax.ml]:
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

File: src/syntax.ml [hunk 16]
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
...
+
+(* Consumes a sequence of table rows (starting with '{tr ...}', which are
+   represented by [`Begin_table_row] tokens).
+
+   This function is called immediately after '{table' ([`Begin_table `Heavy]) is
+   read. The only "valid" way to exit is by reading a [`Right_brace] token,
+   which is consumed. *)
+and heavy_table ~parent_markup ~parent_markup_location input =
+  let rec consume_rows acc ~last_loc =
+    let>> next_token = Reader.until_rbrace input acc in
```
Comment (panglesd) [inline src/syntax.ml]:
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

File: src/syntax.ml [hunk 17]
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
...
+
+(* Consumes a sequence of table cells (starting with '{th ...}' or '{td ... }',
+   which are represented by [`Begin_table_header] [`Begin_table_data] tokens).
+
+   This function is called immediately after '{tr' ([`Begin_table_row]) is
+   read. The only "valid" way to exit is by reading a [`Right_brace] token,
+   which is consumed. *)
+and heavy_table_row ~parent_markup input =
+  let rec consume_cell_items acc =
+    let>> next_token = Reader.until_rbrace input acc in
```
Comment (panglesd) [inline src/syntax.ml]:
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

File: src/syntax.ml [hunk 18]
```diff
@@ -1067,6 +1243,178 @@ and explicit_list_items :
...
+
+(* Consumes a table header (that might start with '{L ...}', '{C ...}' or '{R ... }',
+   which are represented by [`Begin_paragraph_style _] tokens).
+
+   This function is called immediately after '{th' ([`Begin_table_header]) is
+   read. The only "valid" way to exit is by reading a [`Right_brace] token,
+   which is consumed. *)
+and heavy_table_header ~parent_markup input =
+  let rec consume_items acc =
+    let>> next_token = Reader.until_rbrace input acc in
```
Comment (panglesd) [inline src/syntax.ml]:
```suggestion
    Reader.until_rbrace input acc >>> fun next_token ->
```

File: src/syntax.ml [hunk 19]
```diff
@@ -36,6 +38,107 @@ let peek input =
...
+module Table = struct
+  module Light_syntax = struct
+    let valid_align = function
+      | [ { Loc.value = `Word w; _ } ] -> (
+          match String.length w with
+          | 0 -> Some None
+          | 1 -> (
+              match w with
+              | "-" -> Some None
+              | ":" -> Some (Some `Center)
```
Comment (panglesd) [inline src/syntax.ml]: I find it strange to see an `option option`. What is the difference between `Some None` and `None`? Maybe ``Some `Unspecified`` (or a more explicit variant) should replace `Some None`?

File: src/syntax.ml [hunk 20]
```diff
@@ -36,6 +38,107 @@ let peek input =
...
+    let valid_header_row row =
+      List.map
+        (function
+          | `Header (Some x), y -> Some (Some x, y)
+          | `Header None, y -> Some (None, y)
+          | `Data, _ -> None)
+        row
+      |> Option.join_list
+
+    let from_grid grid : Ast.table =
```
Comment (panglesd) [inline src/syntax.ml]: The type of `from_grid` is not `_ grid -> _` (while it is for `Light_syntax.from_grid`), which is confusing...

Anyway, I think if the `cell` type is modified to include the header/normal information, are not that useful anymore.

File: test/test_tables.ml [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+
+    let multiple_headers =
+      test "{table {tr {th}} {tr {th}} {tr {td}}}";
+      [%expect
+        {|
+        ((output
+          (((f.ml (1 0) (1 37))
+            (table (syntax heavy) (header ((cell ())))
+             (data ((row ((cell ()))) (row ((cell ()))))) (align (none))))))
+         (warnings ())) |}]
```
Comment (panglesd) [inline test/test_tables.ml]: `{th ...}` fields should be respected.

File: src/ast.ml [hunk 22]
```diff
@@ -29,6 +30,11 @@ type inline_element =
     text. Similarly the [`Link] constructor has the link itself as first parameter
     and the second is the replacement text. *)
 
+type 'a cell = 'a with_location list
```
Comment (panglesd) [inline src/ast.ml]: Cells should be able to be made header independently of the line:
```suggestion
type 'a cell = 'a with_location list * [ `Header | `Normal ]
```

File: src/syntax.ml [hunk 23]
```diff
@@ -381,6 +484,8 @@ let paragraph : input -> Ast.nestable_block_element with_location =
       [-], has been read, and only whitespace has been read since.
     - [`After_explicit_list_bullet], when a valid explicit bullet, such as [{li],
       has been read, and only whitespace has been read since.
+    - [`After_table_header], when a table header opening markup ('{th') has been read.
+    - [`After_table_cell], when a table cell opening markup ('{td') has been read.
```
Comment (panglesd) [inline src/syntax.ml]: 10 lines above: "The five possible locations" -> "The seven possible locations"

File: src/syntax.ml [hunk 24]
```diff
@@ -594,6 +708,46 @@ let rec block_element_list :
...
+            (Token.describe `Begin_table_row)
+        in
+        Parse_error.not_allowed ~what:(Token.describe token)
+          ~in_what:(Token.describe parent_markup)
+          ~suggestion location
+        |> add_warning input;
+        junk input;
+        consume_block_elements ~parsed_a_tag where_in_line acc
+    (* Bars can never appear directly in block content.
+       They can only appear inside [{t ...}]. *)
```
Comment (panglesd) [inline src/syntax.ml]: It would be nice to be able to write `|` outside of `{t`. Otherwise, that is one more character that we have to escape...

That should be solved in the lexer I guess.

File: src/compat.ml [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+  type 'a t = 'a option = None | Some of 'a
+
+  let is_some = function None -> false | Some _ -> true
+
+  let value_exn = function
+    | None -> failwith "Option.value_exn None"
+    | Some x -> x
+
+  let join_list l =
+    if List.for_all is_some l then Some (List.map value_exn l) else None
```
Comment (panglesd) [inline src/compat.ml]: I'm not that keen on using exceptions... Why not:

```suggestion
  let join_list l =
    let rec loop acc = function
      | [] -> Some (List.rev acc)
      | Some a :: q -> loop (a :: acc) q
      | None :: _ -> None
    in
    loop [] l
```

File: src/ast.ml [hunk 26]
```diff
@@ -29,6 +30,11 @@ type inline_element =
...
+type 'a row = 'a cell list
+type 'a grid = 'a row list
+type 'a abstract_table = 'a row * 'a grid * alignment option list
```
Comment (panglesd) [inline src/ast.ml]: It should be possible for cells to be header independently of the line:
```suggestion
type 'a cell = 'a with_location list * [ `Header | `Normal ]
type 'a row = 'a cell list
type 'a grid = 'a row list
type 'a abstract_table =  'a grid * alignment option list
```

File: src/ast.ml [hunk 27]
```diff
@@ -41,13 +47,18 @@ type nestable_block_element =
...
   | `Math_block of string  (** @since 2.0.0 *) ]
 (** Some block elements may be nested within lists or tags, but not all.
     The [`List] constructor has a parameter of type [\[`Light | `Heavy\]].
     This corresponds to the syntactic constructor used (see the
     {{:https://ocaml.org/releases/4.12/htmlman/ocamldoc.html#sss:ocamldoc-list}manual}).
     *)
 
+and table =
+  [ `Light of inline_element abstract_table
+  | `Heavy of nestable_block_element abstract_table ]
```
Comment (panglesd) [inline src/ast.ml]: I think the information of how the table was given is often irrelevant, so it should be easy to ignore! 
```suggestion
and table = nestable_block_element abstract_table * [ `Light | `Heavy ]
```

**Review by panglesd — CHANGES_REQUESTED** [review]
Nice !

I do have change requested for some things:

- I think that the heavy syntax should allow us to define any cell to be a header cell (just like in html). This allows for several use cases, with two common one being: "vertical" tables (tables where the header is a column) and tables with two "axis" (both a header line and a header column).
Fortunately, I think this will actually simplify the code a lot.
- I believe the lexer should do more work in the case of the light table. Currently, as mentioned by @Julow, opening a forbidden block inside a light table will generate a lot of warnings (where one would have sufficed) and the next closing will be wrongly paired. That's a major issue!
Using a specific lexer for the inside of light tables (which is what is done for verbatim for example) also removes the constraint added by this PR to escape `|` characters outside of tables.
- I don't think the AST type should make such a clear distinction between light and heavy syntax. Similarly to `` `List`, a variant on the side could distinguish between them, the information is almost always irrelevant for consumers of the parser, it has to be easy to ignore.

Another thing: I am not a huge fan of the alignment syntax for heavy tables.
To me, it seems that `{th {R text}}` will only have an effect on the cell, by the nesting. (The fact that it has an effect on the whole columns can only be understood with the light syntax in mind).
Moreover, it prevents possible future uses of `{L }` to specify the alignment on single cell, especially on header cell (if we want that, not sure...).
Instead, we could take inspiration from latex's table (or the `col` html element), where the fact that it acts on columns is evident:
```
{table {col L C R}
  {tr
    {th left-aligned} {th centered} {th right-aligned}}
  }
}
``` 

This leaves room for deciding the line borders in odoc (but maybe that's not something we want):

```
{table {col | L C | R}
  {hline}
  {tr
    {th left-aligned} {th centered} {th right-aligned}}
  }
  {hline}
}
```

File: src/syntax.ml (hunk 14 above)
Comment (gpetiot) [reply src/syntax.ml]: dome

File: src/syntax.ml (hunk 19 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/compat.ml (hunk 25 above)
Comment (gpetiot) [reply src/compat.ml]: done

File: src/ast.ml (hunk 27 above)
Comment (gpetiot) [reply src/ast.ml]: done

File: src/syntax.ml (hunk 23 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: test/test_tables.ml (hunk 21 above)
Comment (gpetiot) [reply test/test_tables.ml]: done

File: src/ast.ml (hunk 22 above)
Comment (gpetiot) [reply src/ast.ml]: done

File: src/ast.ml (hunk 26 above)
Comment (gpetiot) [reply src/ast.ml]: done

File: src/syntax.ml (hunk 20 above)
Comment (gpetiot) [reply src/syntax.ml]: done

File: src/syntax.ml (hunk 24 above)
Comment (gpetiot) [reply src/syntax.ml]: done in [bd19b75](https://github.com/ocaml-doc/odoc-parser/pull/11/commits/bd19b75bd402e199bd9aee1ff4d5b7923cfa18a9)

**Comment by jonludlam** [comment]
I had a quick look to see whether there were any differences running over the set of libraries in odoc's driver with the `ODOC_BENCHMARK` var set to true. I found a two differences - here:

- https://github.com/ocaml/ocaml/blob/trunk/typing/typedtree.mli#L214-L219
- https://github.com/ocaml/ocaml/blob/trunk/middle_end/flambda/unbox_specialised_args.mli#L26-L31

In both instances they are actually code snippets without being enclosed in any sort of verbatim/code tags, but I was slightly surprised nevertheless. It's not a show stopper but I'd be quite interested to know why the differences are there.

**Comment by panglesd** [comment]
Thanks @jonludlam for spotting a bug! It has been fixed in 97fd2a3.

I've updated https://github.com/ocaml/odoc/pull/893 to be in sync with this PR, in case you want to review them as a whole.

File: src/syntax.ml [hunk 28]
```diff
@@ -209,9 +324,11 @@ and delimited_inline_element_list :
     parent_markup:[< Token.t ] ->
     parent_markup_location:Loc.span ->
     requires_leading_whitespace:bool ->
+    context:inline_context ->
```
Comment (jonludlam) [inline src/syntax.ml]: This bit of threaded state is annoying. It's also very similar to the `requires_leading_whitespace` state - I wonder if the two could be combined into a single `Context.t` or similar? I suppose the new bit is relevant to both `delimited_inline_element_list` _and_ `inline_element` so maybe not worth it? It's also possibly worth considering getting rid of the context and just checking in the `light_table` code after we've called this function that the start and end lines are the same?

Comment (jonludlam) [reply src/syntax.ml]: I had a quick go: https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 - I think it's a bit less invasive and keeps the check within the `light_table_row` function. What do you think?

Comment (gpetiot) [reply src/syntax.ml]: It's also my opinion that having the check in `light_table_row` makes the code clearer.

Comment (panglesd) [reply src/syntax.ml]: As a small remark, the location warning in the current PR and https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 will be different:
- In the current PR, it is only the newline. There is one warning per newline.
- In https://github.com/jonludlam/odoc-parser/commit/5c49f986e1094617d566a7a1a20b0106ff954ff3 there is only one warning, with location being the whole inline item.

I think both solutions have different advantages/disadvantages (one is clearer, the other more general), and I don't have a preference. So let's go with @jonludlam's idea!

**Comment by jonludlam** [comment]
Looking good now, thanks @gpetiot!
